### ISSUE TITLE
ACL agent token not used for service registration

### ISSUE THREAD

**Issue by aduzsardi** [hashicorp/consul#4025]
### Description of the Issue (and unexpected/desired result)
With ACL's enabled on the consul cluster with acl_agent_token added in configuration
Service will not be registered on the consul clients/servers, you'll get an error/warning like
```[WARN] agent: Service "nginx" registration blocked by ACLs```

**ACL_AGENT_TOKEN**
`node "" { policy = "write" } service "" { policy = "write" } key "" { policy = "write" }`

If you also add `acl_token ` in the configuration with the same value as `acl_agent_token` or just replace acl_agent_token with acl_token , the service will be registered but makes ACLs useless, because now you can access the UI/API without an ACL being required.

**All consul agents are servers (ie: server: true) in this setup.**

### Reproduction steps
- create consul cluster 
- enable ACLs
- add a alc_master_token and and alc_agent_token to the configs.
- create service definition config file , like
```json
{
    "service": {
        "address": "localhost", 
        "name": "nginx", 
        "port": 80, 
        "tags": [
            "testing"
        ]
    }
}
```
- start consul with `consul agent -config-dir=/etc/config.d/`

> $ ls /etc/consul.d/
> config.json   nginx.json

**nginx.json** posted above

#### config.json
```json
{
    "acl_agent_master_token": "REDACTED", 
    "acl_agent_token": "REDACTED", 
    "acl_datacenter": "dc1", 
    "acl_default_policy": "deny", 
    "acl_down_policy": "extend-cache", 
    "acl_master_token": "REDACTED", 
    "advertise_addr": "192.168.99.100", 
    "bind_addr": "192.168.99.100", 
    "bootstrap_expect": 3, 
    "client_addr": "0.0.0.0", 
    "data_dir": "/opt/consul/data", 
    "datacenter": "dc1", 
    "disable_remote_exec": true, 
    "disable_update_check": true, 
    "enable_script_checks": true, 
    "encrypt": "REDACTED", 
    "encrypt_verify_incoming": true, 
    "encrypt_verify_outgoing": true, 
    "leave_on_terminate": true, 
    "node_name": "node1", 
    "performance": {
        "raft_multiplier": 1
    },  
    "rejoin_after_leave": true, 
    "retry_interval": "15s", 
    "retry_join": [
        "192.168.99.100", 
        "192.168.99.101", 
        "192.168.99.102"
    ], 
    "retry_max": 0, 
    "server": true, 
    "ui": true
}
```

### `consul version` for both Client and Server
Server: `Consul v1.0.6`

### `consul info` for both Client and Server

```
agent:
	check_monitors = 1
	check_ttls = 0
	checks = 1
	services = 4
build:
	prerelease = 
	revision = 9a494b5f
	version = 1.0.6
consul:
	bootstrap = false
	known_datacenters = 1
	leader = false
	leader_addr = 192.168.99.100:8300
	server = true
raft:
	applied_index = 787
	commit_index = 787
	fsm_pending = 0
	last_contact = 51.238765ms
	last_log_index = 787
	last_log_term = 373
	last_snapshot_index = 0
	last_snapshot_term = 0
	latest_configuration = [{Suffrage:Voter ID:1c6c5eaa-62fa-9056-2afc-fa3b3712424c Address:192.168.99.100:8300} {Suffrage:Voter ID:167a30b3-ccd4-8e29-87b4-736d3d0be301 Address:192.168.99.101:8300} {Suffrage:Voter ID:adbb264a-4f36-a358-956c-6c06c9a6e87e Address:192.168.99.102:8300}]
	latest_configuration_index = 1
	num_peers = 2
	protocol_version = 3
	protocol_version_max = 3
	protocol_version_min = 0
	snapshot_version_max = 1
	snapshot_version_min = 0
	state = Follower
	term = 373
runtime:
	arch = amd64
	cpu_count = 48
	goroutines = 82
	max_procs = 2
	os = linux
	version = go1.9.3
serf_lan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 9
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 22
	members = 3
	query_queue = 0
	query_time = 1
serf_wan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 1
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 14
	members = 3
	query_queue = 0
	query_time = 1
```

**Comment by aduzsardi** [comment]
com'on people , nobody has any input on this ?

**Comment by mterron** [comment]
ACL mgmt at the moment is a bit of a mess. There's no cli for them and the 'almost json but not quite' policy you need to craft is rife with error prone escaping sequences. 

Can somebody at Hashicorp assign some priority to dealing with ACLs either via CLI or by having a sane json representation?

**Comment by pearkes** [comment]
You're right @mterron -- lost of room for improvement on ACLs. You can see a wealth of feedback in GitHub issues such as this one.

We plan on picking up a lot of the improvements we made in designing the [Nomad ACL system](https://www.nomadproject.io/guides/acl.html). It is on our roadmap, but not with set timelines.

@aduzsardi unfortunately I don't have time to dig into your specific issue but I hope someone can come back around shortly and give you some feedback. I'm sure you may have already done this but I recommend carefully reviewing the [ACL guide](https://www.consul.io/docs/guides/acl.html).

**Comment by aduzsardi** [comment]
Thank you for the updates and @mterron for the workaround , for now i've used the API to add the services instead of using static config files, that way i can bypass the need for `acl_token` configuration.

**Comment by pearkes** [comment]
Noting that this is a real UX bug. I tagged it and am going to close it due to planned work -- issues like this are useful to reference when designing and implementing the improvements to the ACL system. Because it isn't a cut and dry bug, leaving it open means we may miss it when the upcoming ACL work lands. Please feel free to open a new issue with steps to reproduce an issue if you're seeing an error in the ACL system beyond the confusing design.

**Comment by mterron** [comment]
@pearkes do you know when the ACL work will land and also where the discussion is taking place?

**Comment by pearkes** [comment]
@mterron targeting a release this fall. There's no public discussion (yet) but you can see our general thinking in the acl-cleanup label here on GitHub.
